debian/t1c2pfb.man
.TH "t1c2pfb.pl" "1" "September 2005" "cm-super" "cm-super" 
.PP 
.SH "NAME" 
t1c2pfb \- convert from small .t1c to binary Type1 PFB
.PP 
.SH "SYNOPSIS" 
.PP 
\fBt1c2pfb\fP 
.PP 
.SH "DESCRIPTION" 
.PP 
\fBt1c2pfb\fP extracts fonts saved in a t1c file (compressed version
of type1 files).
.PP 
.SH COPYRIGHT
This program was created by Szab\('o P\('eter and is under public domain.
You may freely use, modify and/or distribute it.
.SH "BUGS" 
.PP 
Problems with this script should be reported to the author or to the
maintainer of this package, <preining@logic\&.at>\&.
.PP 
.SH "AUTHOR" 
.PP 
Written by pts@fazekas\&.hu\&.
.PP 
This manual page was written by Norbert Preining for
the Debian GNU/Linux system\&.  It may be used by other distributions
without contacting the author\&.  Any mistakes or omissions in the
manual page are my fault; inquiries about or corrections to this
manual page should be directed to me (and not to the primary author)\&.


debian/changelog
cm-super (0.3.3-2) unstable; urgency=low

  * list all map files in updmap config file (Closes: #332670)
  * depend only on tex-common, making it impossible to install together
    with teTeX2. Add a README.teTeX2 for backporters. (Closes: #332673)
    This also closes the bug that cm-super-t1.map is left over in
    the updmap.cfg file (Closes: #334177)
  * make font generation less noisy, only mention that it may take
    some time. (Closes: #333745)
  * include the postrm/postinst scripts from lmodern, which are prototypes
    of the Font Package part of the TeX Policy.
  * Install the afm files also to /usr/share/texmf and only make links
    in /usr/X11R6
  * Install defoma hints file
  * The package name is cm-super now, so closes: #325093.

 -- Norbert Preining <preining@logic.at>  Sun, 16 Oct 2005 20:56:49 +0200

cm-super (0.3.3-1) unstable; urgency=low

  * Suggest pfb2t1c2pfb for binary conversion program. In the postinst
    script test first on the t1c2pfb binary and if not present use the
    perl script.
  * changed README.debian, add entry on variants of sharp-s, TODO, copyright
  * fixed old FSF address in copyright file
  * add cm-super-x11 package to make the fonts available from X11
  * removed magic perl header from t1c2pfb.pl
  * created a minimal man page for t1c2pfb (perl script installed in /usr/bin)
  * added lintian overrides
    cm-super: for the md5sum check
    cm-super source: old debian/postinst script is buggy, it is replaced
    cm-super-x11: install into X11R6
  * changed postinst, postrm to work with tetex2
  * Hack around that the pfb files are generated at install time and the
    md5sums are wrong if we ship empty files instead.
  * bump Standards Version to 3.6.2.1 and debhelper compat to 4 
  * add Replaces for pts-tetex-cm-super
  * install map and enc files in the correct locations
  * package renamed and new maintainer

 -- Norbert Preining <preining@logic.at>  Sun,  2 Oct 2005 14:06:00 +0200

pts-tetex-cm-super (0.3.3-4) unstable; urgency=low

  * Full pdftex, dvips and dvipdfm support.

 -- Szabó Péter <pts@fazekas.hu>  Sun, 30 Mar 2003 14:53:28 +0200

pts-tetex-cm-super (0.3.3-3) unstable; urgency=low

  * debian/control updated
  * debain/postinst calls /etc/texmf/dvips/updmap

 -- Szabó Péter <pts@fazekas.hu>  Sun, 30 Mar 2003 14:06:14 +0200

pts-tetex-cm-super (0.3.3-2) unstable; urgency=low

  * Added t2c2pfb.pl

 -- Szabó Péter <pts@fazekas.hu>  Sun, 30 Mar 2003 13:02:24 +0200

pts-tetex-cm-super (0.3.3-1) unstable; urgency=low

  * Initial Debian package (24 Mb).
  * t1c2pfb.c

 -- Szabó Péter <pts@fazekas.hu>  Sat, 29 Mar 2003 22:32:12 +0100

Local variables:
mode: debian-changelog
End:


debian/t1c2pfb.c
#define DUMMY \
set -ex; \
gcc -DNDEBUG=1 -s -O3 -ansi -pedantic \
  -Wall -W -Wstrict-prototypes -Wnested-externs -Winline \
  -Wpointer-arith -Wbad-function-cast -Wcast-qual -Wmissing-prototypes \
  -Wmissing-declarations "$0" -o t1c2pfb; \
exit

/* t1c2pfb.c -- convert from small .t1c (stdin) to binary Type1 PFB (file)
 * by pts@fazekas.hu at Sat Mar 29 18:24:48 CET 2003
 */
#include <stdio.h>
#include <stdlib.h> /* malloc() */
#include <string.h>
#ifdef MSDOS
#  define WIN32 1
#endif
#ifdef WIN32
#  include <fcntl.h>
#  include <io.h>
#endif

#define SEED_EEXEC 55665UL
#define SEED_CHARSTRINGS 4330UL

/** @return buf+len */
static char* eexec_encrypt(char* buf, unsigned long len, unsigned long seed) {
  unsigned char c; /* unsigned is important */
  while (len--!=0) {
    c=*buf^(seed>>8);
    seed=0xFFFFU&((c+seed)*52845U+22719U);
    *buf++=c;
  }
  return buf;
}

#define BLOCK_ASCII '\001'
#define BLOCK_BINARY '\002'
/** Writes a PFB ASCII or binary block */
static void pfb_write(char ty, char* buf, unsigned long len, long pd, FILE *of) {
  char head[6];
  head[0]='\200'; head[1]=ty;
  head[2]=len; head[3]=len>>8; head[4]=len>>16; head[5]=len>>24;
  fwrite(head, 1, 6, of);
  fwrite(buf, 1, len+pd, of);
}

int main(int argc, char** argv) {
  FILE *of;
  int c;
  unsigned long len, k;
  unsigned u;
  char *p, *buf, *bufend, *bufee, *q;
  char t[6];
  char filename[80];
  
  if (argc==3) return 0;
  if (argc==2) {
    if (NULL==(freopen(argv[1], "rb", stdin))) {
      fprintf(stderr, "t1c2pfb: cannot fopen(): %s\n", argv[1]); exit(13);
    }
  }
  #ifdef WIN32
    setmode(0, O_BINARY);
  #endif
  while (0!=(len=fread(t, 1, 6, stdin))) {
    if (6!=len || 0!=memcmp(t, "%!t1c ", 6)) {
      fprintf(stderr, "t1c2pfb: stdin is not a .t1c\n");
      exit(2);
    }
    if (1!=scanf("%lu ", &len)) { fprintf(stderr, "t1c2pfb: incomplete len\n"); exit(3); }
    p=filename; k=sizeof(filename);
    while (1) {
      if (0>(c=getchar())) { fprintf(stderr,"t1c2pfb: incomplete filename\n"); exit(4); }
      if (c=='\n' || c=='\0') break;
      *p++=c;
      if (0==--k) { fprintf(stderr,"t1c2pfb: filename too long\n"); exit(5); }
    }
    *p='\0';
    fprintf(stderr, "extracting filename=(%s) from_len=%lu\n", filename, len);
    if (NULL==(buf=(char*)malloc(len+100))) { fprintf(stderr, "t1c2pfb: out of memory\n"); exit(6); }
    if (len!=fread(buf, 1, len, stdin)) { fprintf(stderr, "t1c2pfb: EOF inside .t1c content\n"); exit(8); }
    strcpy(bufend=buf+len, " -| \ne\n"); /* sentinels */
    remove(filename);
    if (NULL==(of=fopen(filename, "wb"))) { fprintf(stderr, "t1c2pfb: cannot rewrite outfile\n"); exit(9); }

    /* reproduce plain part */
    p=buf; while (p[0]!='\n' || p[1]!='e' || p[2]!='\n') p++;
    if (p>bufend) { fprintf(stderr, "t1c2pfb: incomplete plain part\n"); exit(10); }
    if (p-buf<32) { fprintf(stderr, "t1c2pfb: plain part too short\n"); exit(14); }
    bufee=p+3;
    k=strlen(q="currentfile eexec\n");
    pfb_write(BLOCK_ASCII, buf, p+1-buf+k, -k, of);
    fputs(q, of);
    
    /* Encrypt individual charstrings */
    u=0; p=bufee;
    while (1) {
      c=*p++;
      if (c-'0'+0U<='9'+0U) { u=10*u+c-'0'; }
      else if (c==' ' && p[0]=='-' && p[1]=='|') { /* eexec encoding */
        if (p>bufend) break;
        if (p[2]!=' ') { fprintf(stderr, "t1c2pfb: bad charstring lead\n"); exit(12); }
        #if 0
          fprintf(stderr, "eexec %u\n", u);
        #endif
        #if 0
          p[0]='\236'; p[1]='\264'; p[2]='\365'; p[3]='\225'; /* arbitrary, but encrypted must be binary */
        #endif
        p=eexec_encrypt(p+=3, u, SEED_CHARSTRINGS);
      } else u=0;
    }
    
    /* Encrypt everything after bufee */
    strcpy(bufend, "dup/FontName get exch definefont pop\nmark currentfile closefile\n");
    while (*bufend!='\0') bufend++;
    *--bufee='\0';  *--bufee='\0';  *--bufee='\0';  *--bufee='\0';
    eexec_encrypt(bufee, bufend-bufee, SEED_EEXEC);
    /* ^^^ Imp: dump additional bytes? */
    
    /* dump the results */
    #if 0
      pfb_write(BLOCK_BINARY, bufee, bufend-bufee+532, -532, of);
      /* ^^^ 532 is the length till the end of "cleartomark\n" */
    #else
      pfb_write(BLOCK_BINARY, bufee, bufend-bufee, 0, of);
      fwrite("\200\001\024\002\000\000", 1, 6, of); /* 532 */
    #endif
    for (k=8; k--!=0; ) fputs("0000000000000000000000000000000000000000000000000000000000000000\n", of);
    fputs("cleartomark\n\200\003", of);
    if (ferror(of)) { fprintf(stderr, "t1c2pfb: error writing outfile\n"); exit(11); }
    fclose(of);
    free(buf);
  }
  if (ferror(stdin)) {
    fprintf(stderr, "error reading\n");
    exit(7);
  }
  return 0;
}
